mcp-server/src/utils.ts

```diff
@@ -0,0 +1,173 @@
+/**
+ * @file utils.ts
+ * @description
+ * A utility module for managing an in-memory cache system with TTL (Time-To-Live) support.
+ * This module provides functions to fetch data with caching, track cache statistics, and manage cache entries.
+ * It is designed to optimize performance by reducing redundant network requests and providing insights into cache usage.
+ *
+ * Features:
+ * - In-memory caching with TTL support.
+ * - Cache hit/miss tracking and statistics.
+ * - Automatic cache expiration and cleanup.
+ * - Fetching data with cache fallback.
+ * - Retrieving popular cache entries and memory usage statistics.
+ *
+ * Usage:
+ * - Use `fetchWithCache` to fetch data with caching.
+ * - Use `getCacheStats` to retrieve cache performance metrics.
+ * - Use `clearExpiredCache` or `clearCache` to manage cache entries.
+ * - Use `getPopularCacheEntries` to analyze frequently accessed cache entries.
+ */
+
+
+import { performance } from 'perf_hooks';
+
+interface CacheEntry {
+  value: unknown;
+  expiresAt: number;
+  createdAt: number;
+  hitCount: number;
+}
+
+interface CacheStats {
+  totalEntries: number;
+  totalHits: number;
+  totalMisses: number;
+  hitRate: number;
+  memoryUsage: {
+    heapUsed: number;
+    heapTotal: number;
+    external: number;
+  };
+  oldestEntry?: number | undefined;
+  newestEntry?: number | undefined;
+}
+
+const inMemoryCache: Map<string, CacheEntry> = new Map();
+let cacheHits = 0;
+let cacheMisses = 0;
+
+async function parseResponseBody(response: Response): Promise<unknown> {
+  const contentType = response.headers.get("content-type");
+  if (contentType?.includes("application/json")) {
+    return response.json();
+  }
+  return response.text();
+}
+
+function getCached(url: string): unknown | undefined {
+  const entry = inMemoryCache.get(url);
+  if (!entry) {
+    cacheMisses++;
+    return undefined;
+  }
+  
+  if (Date.now() > entry.expiresAt) {
+    inMemoryCache.delete(url);
+    cacheMisses++;
+    return undefined;
+  }
+  
+  entry.hitCount++;
+  cacheHits++;
+  return entry.value;
+}
+
+function setCached(url: string, value: unknown, ttlSeconds: number): void {
+  const now = Date.now();
+  inMemoryCache.set(url, { 
+    value, 
+    expiresAt: now + ttlSeconds * 1000,
+    createdAt: now,
+    hitCount: 0
+  });
+}
+
+export async function fetchWithCache(url: string, ttlSeconds: number): Promise<unknown> {
+  const startTime = performance.now();
+  
+  try {
+    const cached = getCached(url);
+    if (cached !== undefined) {
+      console.error(`Cache hit for ${url} (${(performance.now() - startTime).toFixed(2)}ms)`);
```

Comment:
Why use the console error log type here?

Reply:
Good catch! And sorry, I forgot to update that 😅 
I used it for debugging purposes (when I was testing it on my local), but I've updated that to,

```js
console.log(`Cache hit for ${url} (${(performance.now() - startTime).toFixed(2)}ms)`);
```

So it should be fine now! 😄